### PR TITLE
Merge serialization revamp

### PR DESCRIPTION
**Pull request by neuneck**
I still have to fix the tests, but this is how I imagine the functionality to look like and wanted to get this out before the weekend.

### CODE REVIEW

**Comment by neuneck** [comment]
Note: I realized that `FunctionNode` can not be pickled generally and nodes created with the `@Node` decorator can never be pickled. The reason is that their evaluation function is not defined at the top level of of a module and pickle can not handle closures.

Therefore, I added a `to_pickle` function to the `FunctionNode` class and had it raise a `NotImplementedError` with appropriate message.

See https://docs.python.org/3.7/library/pickle.html#pickle-picklable

File: flowpipe/graph.py [hunk 1]
```diff
@@ -209,18 +212,52 @@ def upstream_ready(processes, node):
...
+
+    def serialize(self):  # pragma: no cover
+        """Serialize the graph in it's grid form.
+
+        Deprecated.
+        """
+        trace = inspect.getouterframes(inspect.currentframe())
+        if not any(frame.function == 'to_json' for frame in trace):
+            warnings.warn(
+                'Graph.serialize is deprecated. Use Graph.to_json instead',
```
Comment (PaulSchweizer) [inline flowpipe/graph.py]: Maybe better: "Use Graph.to_json **or Graph.to_pickle** instead"?

File: flowpipe/graph.py [hunk 2]
```diff
@@ -209,18 +212,52 @@ def upstream_ready(processes, node):
...
+    @staticmethod
+    def from_json(data):
+        """De-serialize from the given json data."""
+        return deserialize_graph(data)
+
+    @staticmethod
+    def deserialize(data):  # pragma: no cover
         """De-serialize from the given json data."""
+        warnings.warn(
+            'Graph.deserialize is deprecated. Use Graph.from_json instead',
```
Comment (PaulSchweizer) [inline flowpipe/graph.py]: ... or Graph.from_pickle ...

File: flowpipe/node.py [hunk 3]
```diff
@@ -170,8 +171,29 @@ def on_input_plug_set_dirty(self):
...
         """Serialize the node to json."""
+        return self._serialize()
+
+    def serialize(self):  # pragma: no cover
+        """Serialize the node to json.
+
+        Deprecated and kept for backwards compatibility.
+        """
+        trace = inspect.getouterframes(inspect.currentframe())
+        if not any(frame.function == 'to_json' for frame in trace):
```
Comment (PaulSchweizer) [inline flowpipe/node.py]: I'm not quite sure why we'd need this check?

File: flowpipe/node.py [hunk 4]
```diff
@@ -473,6 +508,11 @@ def _initialize(self, func, outputs, metadata):
             for output in outputs:
                 OutputPlug(output, self)
 
+    def to_pickle(self):  # pragma: no cover
+        """Pickle the node. -- DOES NOT WORK FOR FunctionNode!"""
+        raise NotImplementedError("Pickling is not implemented for "
```
Comment (PaulSchweizer) [inline flowpipe/node.py]: Maybe we should add something like: "If you want to use pickle with this node, consider converting your Node to a class subclassing directly from flowpipe.node.INode"

**Comment by neuneck** [comment]
@PaulSchweizer Very good points. I adapted the messages accordingly.

**Comment by neuneck** [comment]
If things are good with the serialization now, we could merge and create a new release. Or is there something left open?

**Pull request merged**